backend/file-orchestrator-service/src/jvmMain/kotlin/service/SyncFolderService.kt

```diff
@@ -0,0 +1,320 @@
+                    f.path like any((select unnest(:parentIds::text[])))
+            """
+        ).rows.map { row ->
+            val folderId = row.getLong("resource") ?: 0
+            val folderPath = row.getString("path") !!
+            val resource = batch.items.first {
+                it.resource.id == extractPathMetadata(folderPath).collection
+            }.added;
+
+            val newSyncType = if (resource.any { it.permissions.contains(Permission.EDIT) }) {
+                SynchronizationType.SEND_RECEIVE
+            } else if (resource.any { it.permissions.contains(Permission.READ)}) {
+                SynchronizationType.SEND_ONLY
+            } else {
+                null
+            }
+
+            Pair(folderId, newSyncType)
+        }
+
+        if (affectedFolders.isEmpty()) {
+            return
+        }
+
+        session.sendPreparedStatement(
+            {
+                setParameter("ids", affectedFolders.filter { it.second == null }.map { it.first })
+            },
+            """
+                delete from file_orchestrator.sync_folders
+                where resource in (select unnest(:ids::bigint[]))
+            """
+        )
+
+        session.sendPreparedStatement(
+            {
+                setParameter("ids", affectedFolders.filter { it.second != null }.map { it.first })
+                setParameter("sync_types", affectedFolders.filter { it.second != null }.map { it.second.toString() })
+            },
+            """
+                update file_orchestrator.sync_folders set
+                    sync_type = updates.type
+                from (select unnest(:ids::bigint[]), unnest(:sync_types::text[])) updates(id, type)
+                where resource = updates.id
+            """
+        )
+
+        proxy.bulkProxy(
+            ActorAndProject(Actor.System, null),
+            BulkRequest(affectedFolders),
+            BulkProxyInstructions.pureProcedure(
+                service = this,
+                retrieveCall = { providerApi(it).onFilePermissionsUpdated },
+                requestToId = { it.first.toString() },
+                resourceToRequest = { req, res ->
+                    res.copy(status = res.status.copy(syncType = req.second))
+                }
+            )
+        )
+    }
+
+    private suspend fun onFilesMoved(batch: List<FilesMoveRequestItem>) {
+        removeSyncFolders(batch.map { it.oldId })
+    }
+
+    private suspend fun onFilesDeleted(request: List<FindByStringId>) {
+        removeSyncFolders(request.map { it.id })
+    }
+
+    private suspend fun removeSyncFolders(paths: List<String>) {
+        db.withSession { session ->
+            val affectedFolders = session.sendPreparedStatement(
+                {
+                    setParameter("ids", paths)
+                    setParameter("parentIds", paths.map { "$it/%" })
+                },
+                """
+                        select * from provider.resource r
```

Comment:
You should be able to use the `to_json` SQL function, which you have already implemented.